src/raft.rs

```diff
@@ -235,7 +235,9 @@ impl<T: Storage> Raft<T> {
     pub fn new(c: &Config, store: T, logger: &Logger) -> Result<Self> {
         c.validate()?;
         let logger = logger.new(o!("raft_id" => c.id));
-        let raft_state = store.initial_state()?;
+        let mut raft_state = store.initial_state()?;
+        // If commit index is smaller than applied index, forward it.
+        raft_state.hard_state.commit = cmp::max(raft_state.hard_state.commit, c.applied);
```

Comment:
I don't think it's appropriate to add the assumption. It can break consistency silently if users don't write implementation correctly. Users should guarantee **_only applying committed logs_**. Hence, they should ensure the invariant `applied <= commit`.

Reply:
> the hard state may be persisted after applying some entries

If users try to do that, then they should be responsible to give a correct commit index. For example, recording the dependent commit index and commit term with the apply index.

> ...raft-rs must depend on some behaviors outside to guarantee correctness.

I don't think raft-rs depends on apply index > commit index.

Reply:
> If users try to do that, then they should be responsible to give a correct commit index. For example, recording the dependent commit index and commit term with the apply index.

For async ready, commit index may be forwarded after calling `on_persist_last_ready`.  Then the result has committed entries but no new hard state. It's redundant to put hard state in `PersistLastReadyResult` because the only change is `commit` variable.

> I don't think raft-rs depends on apply index > commit index.
Yes. raft-rs does not depend on applied index > commit index.
But it depends on the fact that users must apply the committed entries, which must be satisfied.
If so, it's correct to forward the commit index to applied index during initialization.

Reply:
> ...It's redundant to put hard state in PersistLastReadyResult because the only change is commit variable.

It's not. Unless you move the commit out of hard state, otherwise it should be updated.

> ...it depends on the fact that users must apply the committed entries, which must be satisfied...

Indeed, so we have to check `commit index >= apply index`.

> ...If so, it's correct to forward the commit index to applied index during initialization.

The problem is only users know if it's so. You can't just assume everyone implement it in this way.

Reply:
We can check `commit-index >= apply-index` in `raft-rs` while change the `commit-index` of `hard-state`  in `initial_state`, which is implemented in TiKV.